src/dmd/e2ir.d

```diff
@@ -5266,6 +5266,9 @@ elem *callfunc(const ref Loc loc,
     elem *eside = null;
     elem *eresult = ehidden;
 
+    if (fd && fd.flags & FUNCFLAG.compileTimeOnly)
+        fd.error("is annotated with `pragma(ctfe)` and may not be called at runtime");
```

Comment:
Ideally this check would be somewhere else otherwise consumers of the frontend have to implement it themselves. Whether they want to generate object code is up to them but they shouldn't have to reimplement error messages.